testing/unstable_snapshot.ts

```diff
@@ -0,0 +1,411 @@
+   * of snapshots updated after all tests have run.
+   *
+   * This method can safely be called more than once and will only register the teardown
+   * function once in a context.
+   */
+  registerTeardown() {
+    if (!this.#teardownRegistered) {
+      globalThis.addEventListener("unload", this.#teardown);
+      this.#teardownRegistered = true;
+    }
+  }
+
+  /**
+   * Creates a snapshot by index. Updates will be written to the test file when all
+   * tests have run.
+   */
+  createSnapshot(request: SnapshotUpdateRequest, format: boolean | undefined) {
+    this
+      .#lineColumnToSnapshot[`${request.lineNumber}:${request.columnNumber}`] =
+        request.actualSnapshot;
+
+    if (format === undefined) format = true;
+
+    if (this.#format === undefined) {
+      this.#format = format;
+    } else if (this.#format !== format) {
+      this.#format = "error";
+    }
+  }
+}
+
+/**
+ * Make an assertion that `actual` matches `expectedSnapshot`. If they do not match,
+ * then throw.
+ *
+ * Type parameter can be specified to ensure values under comparison have the same type.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @example Usage
+ * ```ts no-assert
+ * import { assertInlineSnapshot } from "@std/testing/unstable-snapshot";
+ *
+ * Deno.test("snapshot", (t) => {
+ *   assertInlineSnapshot<number>(t, 2, `2`);
+ * });
+ * ```
+ * @typeParam T The type of the snapshot
+ * @param context The test context
+ * @param actual The actual value to compare
+ * @param expectedSnapshot The expected snapshot, or \`CREATE\` to create
+ * @param options The options
+ */
+export function assertInlineSnapshot<T>(
+  context: Deno.TestContext,
+  actual: T,
+  expectedSnapshot: string,
+  options?: InlineSnapshotOptions<T>,
+): void;
+/**
+ * Make an assertion that `actual` matches `expectedSnapshot`. If they do not match,
+ * then throw.
+ *
+ * Type parameter can be specified to ensure values under comparison have the same type.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @example Usage
+ * ```ts no-assert
+ * import { assertInlineSnapshot } from "@std/testing/unstable-snapshot";
+ *
+ * Deno.test("snapshot", (t) => {
+ *   assertInlineSnapshot<number>(t, 2, `2`);
+ * });
+ * ```
+ * @typeParam T The type of the snapshot
+ * @param context The test context
+ * @param actual The actual value to compare
+ * @param expectedSnapshot The expected snapshot, or \`CREATE\` to create
+ * @param message The optional assertion message
+ */
+export function assertInlineSnapshot<T>(
+  context: Deno.TestContext,
+  actual: T,
+  expectedSnapshot: string,
+  message?: string,
+): void;
+export function assertInlineSnapshot(
+  context: Deno.TestContext,
+  actual: unknown,
+  expectedSnapshot: string,
+  msgOrOpts?: string | InlineSnapshotOptions<unknown>,
+): void {
+  const options = getOptions(msgOrOpts);
+  const assertInlineSnapshotContext = AssertInlineSnapshotContext.fromContext(
+    context,
+  );
```

Comment:
I might be missing something in the discussion, but can we get the file name from the CallSite object as @stefnotch's suggested implementation does? I feel that'd be more elegant design. What do you think?